File: test-suite/src/test/java/io/micronaut/gcp/pubsub/subscriber/ReactiveSubscriberTest.java
package io.micronaut.gcp.pubsub.subscriber;

import com.google.pubsub.v1.PubsubMessage;
import io.micronaut.context.annotation.Property;
import io.micronaut.context.annotation.Requires;
import io.micronaut.gcp.pubsub.annotation.PubSubClient;
import io.micronaut.gcp.pubsub.annotation.Topic;
import io.micronaut.gcp.pubsub.support.Animal;
import io.micronaut.http.MediaType;
import io.micronaut.test.annotation.MockBean;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import reactor.core.publisher.Mono;

import java.nio.charset.StandardCharsets;
import java.time.Duration;

import static org.awaitility.Awaitility.await;

@MicronautTest
@Property(name = "spec.name", value = "ReactiveSubscriberTest")
public class ReactiveSubscriberTest {

    @Inject
    TestPublisher publisher;

    Object unwrappedResult;

    @BeforeEach
    void setup() {
        unwrappedResult = null;
    }

    @Test
    void testRawBytes() {
        byte[] bytesSent = "foo".getBytes(StandardCharsets.UTF_8);
        publisher.publishRaw(bytesSent);

        await().atMost(Duration.ofSeconds(2)).until(() -> unwrappedResult != null);
        Assertions.assertInstanceOf(byte[].class, unwrappedResult);
        byte[] convertedResult = (byte[]) unwrappedResult;
        String decodedMessage = new String(convertedResult, StandardCharsets.UTF_8);
        Assertions.assertEquals("foo", decodedMessage);
    }

    @Test
    void testNativeMessage() {
        byte[] bytesSent = "foo".getBytes(StandardCharsets.UTF_8);
        publisher.publishNative(bytesSent);

        await().atMost(Duration.ofSeconds(2)).until(() -> unwrappedResult != null);
        Assertions.assertInstanceOf(PubsubMessage.class, unwrappedResult);
        PubsubMessage convertedResult = (PubsubMessage) unwrappedResult;
        String decodedMessage = convertedResult.getData().toString(StandardCharsets.UTF_8);
        Assertions.assertEquals("foo", decodedMessage);
    }

    @Test
    void testJsonPojo() {
        Animal dog = new Animal("dog");
        publisher.publishAnimal(dog);

        await().atMost(Duration.ofSeconds(2)).until(() -> unwrappedResult != null);
        Assertions.assertInstanceOf(Animal.class, unwrappedResult);
        Animal convertedResult = (Animal) unwrappedResult;
        Assertions.assertEquals("dog", convertedResult.getName());
    }

    @Test
    void testXmlPojo() {
        Animal dog = new Animal("cat");
        publisher.publishAnimalAsXml(dog);

        await().atMost(Duration.ofSeconds(2)).until(() -> unwrappedResult != null);
        Assertions.assertInstanceOf(Animal.class, unwrappedResult);
        Animal convertedResult = (Animal) unwrappedResult;
        Assertions.assertEquals("cat", convertedResult.getName());
    }

    @MockBean(MessageProcessor.class)
    MessageProcessor mockMessageProcessor() {
        return new MessageProcessor() {
            @Override
            public Mono<Boolean> handleByteArrayMessage(byte[] message) {
                unwrappedResult = message;
                return MessageProcessor.super.handleByteArrayMessage(message);
            }

            @Override
            public Mono<Boolean> handlePubSubMessage(PubsubMessage pubsubMessage) {
                unwrappedResult = pubsubMessage;
                return MessageProcessor.super.handlePubSubMessage(pubsubMessage);
            }

            @Override
            public Mono<Boolean> handleAnimalMessage(Animal message) {
                unwrappedResult = message;
                return MessageProcessor.super.handleAnimalMessage(message);
            }
        };
    }

    @Singleton
    @PubSubClient
    @Requires(property = "spec.name", value = "ReactiveSubscriberTest")
    interface TestPublisher {
        @Topic("raw-subscription") void publishRaw(byte[] payload);
        @Topic("native-subscription") void publishNative(byte[] payload);
        @Topic("animals") void publishAnimal(Animal animal);
        @Topic(value = "animals-legacy", contentType = MediaType.APPLICATION_XML) void publishAnimalAsXml(Animal animal);
    }
}

File: test-suite-python/src/test/python/micronaut/gcp/pubsub/subscriber/ReactiveSubscriberTest.py
from abc import ABC, abstractmethod
from typing import Annotated

import java
from com.google.pubsub.v1 import PubsubMessage
from jakarta.inject import Inject
from java.util.concurrent import TimeUnit
from micronaut.context.annotation import Property, Requires
from micronaut.gcp.pubsub.annotation import PubSubClient, Topic
from micronaut.gcp.pubsub.support.Animal import Animal
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import BeforeEach, Disabled, Test

from micronaut.gcp.pubsub.subscriber.MessageProcessor import MessageProcessor


@PubSubClient
@Requires(property="spec.name", value="ReactiveSubscriberTest")
class ReactiveTestPublisher(ABC):

    @Topic("raw-subscription")
    @abstractmethod
    def publish_raw(self, payload: bytes) -> None:
        ...

    @Topic("native-subscription")
    @abstractmethod
    def publish_native(self, payload: bytes) -> None:
        ...

    @Topic("animals")
    @abstractmethod
    def publish_animal(self, animal: Animal) -> None:
        ...

    @Topic(value="animals-legacy", contentType="application/xml")
    @abstractmethod
    def publish_animal_as_xml(self, animal: Animal) -> None:
        ...


@Property(name="spec.name", value="ReactiveSubscriberTest")
@MicronautTest
class ReactiveSubscriberTest:
    publisher: Annotated[ReactiveTestPublisher, Inject]
    messageProcessor: Annotated[MessageProcessor, Inject]

    @BeforeEach
    def setup(self):
        self.messageProcessor.unwrapped_result = None

    def await_result(self):
        for _ in range(40):
            if self.messageProcessor.unwrapped_result is not None:
                return self.messageProcessor.unwrapped_result
            TimeUnit.MILLISECONDS.sleep(50)
        assert self.messageProcessor.unwrapped_result is not None, "no message received"

    @Disabled("TODO(python): bytes as a generic type argument (Mono[bytes]) is compiled to Mono<Byte>, so the message body binder cannot resolve byte[] as the body type")
    @Test
    def test_raw_bytes(self):
        self.publisher.publish_raw("foo".encode())

        result = self.await_result()
        assert bytes(result).decode() == "foo"

    @Test
    def test_native_message(self):
        self.publisher.publish_native("foo".encode())

        result = self.await_result()
        assert java.instanceof(result, PubsubMessage)
        assert result.getData().toStringUtf8() == "foo"

    @Test
    def test_json_pojo(self):
        self.publisher.publish_animal(Animal("dog"))

        result = self.await_result()
        assert result.name == "dog"

    @Test
    def test_xml_pojo(self):
        self.publisher.publish_animal_as_xml(Animal("cat"))

        result = self.await_result()
        assert result.name == "cat"
